functions/go/set-image/transformer/images_transformer.go

```diff
@@ -0,0 +1,170 @@
+			t.Image.Name = val
+		case "newName":
+			t.Image.NewName = val
+		case "newTag":
+			t.Image.NewTag = val
+		case "digest":
+			t.Image.Digest = val
+		default:
+			return fmt.Errorf("ConfigMap has wrong field name %v", key)
+		}
+	}
+	return nil
+}
+
+// validateInput validates the inputs passed into via the functionConfig
+func (t *SetImage) validateInput() error {
+	// TODO: support container name and only one argument input in the next PR
+	if t.Image.Name == "" {
+		return fmt.Errorf("must specify `name`")
+	}
+	if t.Image.NewName == "" && t.Image.NewTag == "" && t.Image.Digest == "" {
+		return fmt.Errorf("must specify one of `newName`, `newTag`, or `digest`")
+	}
+	if t.Image.NewTag != "" && t.Image.Digest != "" {
+		return fmt.Errorf("image `newTag` and `digest` both set, set only one")
+	}
+	return nil
+}
+
+func (t *SetImage) setPodSpecContainers(o *fn.KubeObject) error {
+	podSpec := o.GetMap("spec").GetMap("template").GetMap("spec")
+	for _, vecObj := range podSpec.GetSlice("containers") {
+		if err := t.updateImages(vecObj, &t.Image, o); err != nil {
+			return err
+		}
+	}
+	for _, vecObj := range podSpec.GetSlice("iniContainers") {
+		if err := t.updateImages(vecObj, &t.Image, o); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *SetImage) setPodContainers(o *fn.KubeObject) error {
+	spec := o.GetMap("spec")
+	for _, vecObj := range spec.GetSlice("containers") {
+		if err := t.updateImages(vecObj, &t.Image, o); err != nil {
+			return err
+		}
+	}
+	for _, vecObj := range spec.GetSlice("iniContainers") {
+		if err := t.updateImages(vecObj, &t.Image, o); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *SetImage) hasPodSpecContainers(o *fn.KubeObject) bool {
+	spec := o.GetMap("spec")
+	if spec == nil {
+		return false
+	}
+	template := spec.GetMap("template")
+	if template == nil {
+		return false
+	}
+	podSpec := template.GetMap("spec")
+	return podSpec != nil
+}
+
+func (t *SetImage) hasPodContainers(o *fn.KubeObject) bool {
+	spec := o.GetMap("spec")
+	return spec != nil
+}
+
+// getNewImageName return the new name for image field
+func getNewImageName(oldValue string, newImage *image_util.Image) string {
+	name, tag := image_util.Split(oldValue)
+	if newImage.NewName != "" {
+		name = newImage.NewName
+	}
+	if newImage.NewTag != "" {
+		tag = ":" + newImage.NewTag
+	}
+	if newImage.Digest != "" {
+		tag = "@" + newImage.Digest
+	}
+	newName := name + tag
+	return newName
+}
+
+// updateImages update the image for a given fieldpath
+func (t *SetImage) updateImages(o *fn.SubObject, newImage *image_util.Image, parentO *fn.KubeObject) error {
+	oldValue := o.NestedStringOrDie("image")
+	if !image_util.IsImageMatched(oldValue, newImage.Name) {
+		return nil
+	}
+	newName := getNewImageName(oldValue, newImage)
+	err := o.SetNestedString(newName, "image")
+
+	msg := fmt.Sprintf("updated image from %v to %v", oldValue, newName)
+	t.context.ResultInfo(msg, parentO)
+	t.resultCount += 1
+	return err
```

Comment:
It's better to log result (info, error or warning) in the same function, otherwise it would be harder to figure out what each function does. 

Also, passing the parentO into the function merely for the `ResultInfo` seems to be an overkill. See previous comment https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/904#discussion_r946248335

Reply:
Some pseudo code to explain my comments (and other comments about the duplicate code issue around "container" "initContainers", unnecessary argument passed into the function) 
 
```go
func (t *SetImage) updatePodImages(pod *fn.SubObject, ctx *fn.Context) error {
       var containers fn.SliceSubObjects 
       containers = append(containers, pod.GetSlice("initContainers")...)
       containers = append(containers, pod.GetSlice("containers")...)

       for _, container := range containers {
         oldName := o.NestedStringOrDie("image")
         if !IsImageMatched(oldName, t.Image.Name) {
             continue
         }
         newName := getNewImageName(oldName, newImage)
         if newImageName == oldName {
            continue
         }
         err := o.SetNestedString(newName, "image")
         if err != nil {
             ctx.ResultErr()
         }  else {
             ctx.ResultInfo
         }
       }
```

Reply:
For one Object, it can have multiple changes in its images. If I just return a bool, it is not enough to record all the old names that have changed. One way is to keep an internal struct, that tracks all the results, and log them in the end. Another way is to pass in its parent Object, and then I can have the logging right after the update

Reply:
Sorry somehow I missed your previous comments. Will take a look now.

Reply:
no return error should be needed. The code is pseudo code

Reply:
I know it is pseudo code and I am not asking about return error. My reply was edited. Do you mind checking my updated comments?

Reply:
We talked about that the "image" refers to the "Pod" container image. I didn't read the previous code carefully enough, I think the code does not show this aspect. I'd expect 
1. A "Pod" resource is determined by its "Kind" not the existence of "spec.template.spec". Specifically, the `hasPodContainers` function which checks the existence of `spec` is not right and misleading because all resources have `spec` field, but they are not guaranteed to be a Pod.   
2. Can you verify that in core k8s resources (the same approach as we discussed in set-labels) or kustomize setImage (Maybe @natasha41575 can help giving some guidance here) that only Pod should have containers that need to set the image. 
3. For `SetImages` functions, a more reasonable code structure should be:  for any resources other than Pod, if it contains pod object, get its `Pod` SubObject and then call the Pod `SetContainers` functions.
4. `SetPodImages` --> maybe "ContainerImage" is a more accurate name, at least that's how the code should be structured.

Reply:
I have verified and code is updated. I checked the structs in kubernetes, it seems they will not contain `Pod` as `SubObject`